dev/ProgressRing/ProgressRing.cpp

```diff
@@ -0,0 +1,119 @@
+#include "pch.h"
+#include "common.h"
+#include "ProgressRing.h"
+#include "RuntimeProfiler.h"
+#include "ResourceAccessor.h"
+#include "math.h"
+
+ProgressRing::ProgressRing()
+{
+    __RP_Marker_ClassById(RuntimeProfiler::ProfId_ProgressRing);
+
+    SetDefaultStyleKey(this);
+
+    RegisterPropertyChangedCallback(winrt::RangeBase::ValueProperty(), { this, &ProgressRing::OnRangeBasePropertyChanged });
+
+    SizeChanged({ this, &ProgressRing::OnSizeChanged });
+}
+
+void ProgressRing::OnApplyTemplate()
+{
+    winrt::IControlProtected controlProtected{ *this };
+
+    m_layoutRoot.set(GetTemplateChildT<winrt::Grid>(s_LayoutRootName, controlProtected));
+    m_outlineFigure.set(GetTemplateChildT<winrt::Windows::UI::Xaml::Media::PathFigure>(s_OutlineFigureName, controlProtected));
+    m_outlineArc.set(GetTemplateChildT<winrt::Windows::UI::Xaml::Media::ArcSegment>(s_OutlineArcName, controlProtected));
+    m_barFigure.set(GetTemplateChildT<winrt::Windows::UI::Xaml::Media::PathFigure>(s_BarFigureName, controlProtected));
+    m_barArc.set(GetTemplateChildT<winrt::Windows::UI::Xaml::Media::ArcSegment>(s_BarArcName, controlProtected));
+
+    RenderAll();
+}
+
+void  ProgressRing::OnPropertyChanged(const winrt::DependencyPropertyChangedEventArgs& args)
+{
+    winrt::IDependencyProperty property = args.Property();
+    
+    // TODO: Implement
+}
+
+void ProgressRing::OnSizeChanged(const winrt::IInspectable&, const winrt::IInspectable&)
+{
+    RenderAll();
+}
+
+void ProgressRing::OnRangeBasePropertyChanged(const winrt::DependencyObject& sender, const winrt::DependencyProperty& args)
+{
+    RenderSegment();
+}
+
+void ProgressRing::OnStrokeThicknessPropertyChanged(const winrt::DependencyPropertyChangedEventArgs& args)
+{
+    RenderAll();
+}
+
+winrt::Windows::Foundation::Size ProgressRing::ComputeEllipseSize(const double thickness)
+{
+    const double safeThickness = std::max(thickness, static_cast<double>(0.0));
+    const double actualWidth = ProgressRing::Width();
+    const double width = std::max((ProgressRing::Width() - safeThickness) / 2.0, 0.0);
+    const double height = std::max((ProgressRing::Height() - safeThickness) / 2.0, 0.0);
+
+    return winrt::Windows::Foundation::Size(static_cast<float>(width), static_cast<float>(height));
+}
+
+
+void ProgressRing::RenderSegment()
+{
+    if (auto&& progressRing = m_layoutRoot.get())
+    {
+        auto&& barFigure = m_barFigure.get();
+        auto&& barArc = m_barArc.get();
+        const double thickness = ProgressRing::StrokeThickness();
+        const double maximum = Maximum();
+        const double minimum = Minimum();
+
+        const double range = maximum - minimum;
+        const double delta = Value() - minimum;
+
+        double normalizedRange = (range == 0.0) ? 0.0 : (delta / range);
+        normalizedRange = std::min(std::max(0.0, normalizedRange), 0.9999);
+
+        const double angle = 2 * M_PI * normalizedRange;
+        const auto size = ComputeEllipseSize(thickness);
```

Comment:
>ComputeEllipseSize [](start = 26, length = 18)

The only time you call ComputeEllispeSize is with the thickness parameter set to StrokeThickness(). I think you are doing this because you don't want the helper and the parent to both have to call the DependencyProperty getter. @JevanSaks is the cost of the DP getter large enough that redundant calls should be avoided? I think the cost is pretty low in which case I would just call the getter in the helper.  If the cost is too high I think better would be to have a member variable `m_strokeThickness` which is updated in OnStrokeThicknessPropertyChanged to the DP value, then you can reference this member variable whenever you need it.